COLLABORATOR_GUIDE.md

```diff
@@ -84,6 +87,197 @@ All pull requests that modify executable code should be subjected to
 continuous integration tests on the
 [project CI server](https://ci.nodejs.org/).
 
+### Internal vs. Public API
+
+Due to the nature of the JavaScript language, it can often be difficult to
+establish a clear distinction between which parts of the Node.js implementation
+represent the "public" API Node.js users should assume to be stable and which
+are considered part of the "internal" implementation detail of Node.js itself.
+A general rule of thumb has been to base the determination off what
+functionality is actually *documented* in the official Node.js API
+documentation. However, it has been repeatedly demonstrated that either the
+documentation does not completely cover implemented behavior or that Node.js
+users have come to rely heavily on undocumented aspects of the Node.js
+implementation.
+
+While there are numerous exceptions, the following general rules should be
+followed to determine which aspects of the Node.js API are considered
+"internal":
+
+- Any and all functionality exposed via `process.binding(...)` is considered to
+  be internal and *not* part of the Node.js Public API.
+- Any and all functionality implemented in `lib/internal/**/*.js` that is not
+  re-exported by code in `lib/*.js`, or is not documented as part of the
+  Node.js Public API, is considered to be internal.
+- Any object property or method whose key is a non-exported `Symbol` is
+  considered to be an internal property.
+- Any object property or method whose key begins with the underscore `_` prefix,
+  and is not documented as part of the Node.js Public API, is considered to be
+  an internal property.
+- Any object, property, method, argument, or event not documented in the
+  Node.js documentation is considered to be internal.
+- Any native C/C++ APIs/ABIs exported by the Node.js `*.h` header files that
+  are hidden behind the `NODE_WANT_INTERNALS` flag are considered to be
+  internal.
+
```

Comment:
Also, cf. https://github.com/nodejs/node/pull/10866#issuecomment-275792153, we need to state more about what *is* the API: error message strings, throwing errors on previously ignored invalid input.